File: produtos/src/main/java/br/com/bota/service/impl/ProdutoServiceImpl.java
package br.com.bota.service.impl;

import br.com.bota.entity.Produto;
import br.com.bota.repository.ProdutoRepository;
import br.com.bota.service.ProdutoService;
import br.com.bota.service.rule.ProdutoNotFoundException;
import lombok.extern.slf4j.Slf4j;

import javax.inject.Inject;
import javax.transaction.Transactional;
import java.util.List;
import java.util.UUID;

@Slf4j
public class ProdutoServiceImpl implements ProdutoService {
    @Inject
    private ProdutoRepository repository;

    @Transactional
    @Override
    public Produto salvar(Produto product) {
        if (product.getId() == null) {
            return repository.save(product);
        } else {
            var existente = this.procurarPorId(product.getId());
            existente.setDescricao(product.getDescricao());
            existente.setQuantidade(product.getQuantidade());
            existente.setValor(product.getValor());
            return repository.update(existente);
        }
    }

    @Transactional
    public void baixarEstoque(UUID id, Integer quantidade) {
        var produto = procurarPorId(id);
        log.info("Antes de baixar o estoque: {}", produto);
        produto.setQuantidade(produto.getQuantidade() - quantidade);
        log.info("Depois de baixar o estoque: {}", produto);
        repository.update(produto);
    }

    @Override
    public Produto procurarPorId(UUID id) {
        return repository.findById(id).orElseThrow(() -> new ProdutoNotFoundException());
    }

    @Override
    public Boolean verificarQuantidade(UUID id) {
        return this.procurarPorId(id).getQuantidade() > 0;
    }

    @Override
    public List<Produto> procurar() {
        return repository.findAll();
    }
}

File: loja-kotlin/kprodutos/src/main/kotlin/br/com/bota/service/impl/ProdutoServiceImpl.kt
package br.com.bota.service.impl

import br.com.bota.entity.Produto
import br.com.bota.exceptions.ProdutoNotFoundException
import br.com.bota.repository.ProdutoRepository
import br.com.bota.service.ProdutoService
import jakarta.inject.Inject
import jakarta.inject.Singleton
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*
import javax.transaction.Transactional

@Singleton
open class ProdutoServiceImpl @Inject constructor(private val repository: ProdutoRepository) : ProdutoService {
    private val log: Logger = LoggerFactory.getLogger(ProdutoServiceImpl::class.java)

    @Transactional
    override fun salvar(produto: Produto): Produto {
        return if (produto.id == null) {
            repository.save(produto)
        } else {
            val existente = this.procurarPorId(produto.id!!)
            existente.descricao = produto.descricao
            existente.quantidade = produto.quantidade
            existente.valor = produto.valor
            repository.update(existente)
        }
    }

    @Transactional
    override fun baixarEstoque(id: UUID, quantidade: Int) {
        val produto = procurarPorId(id)
        log.info("Antes de baixar o estoque: {}", produto)
        produto.quantidade = produto.quantidade?.minus(quantidade)
        log.info("Depois de baixar o estoque: {}", produto)
        repository.update(produto)
    }

    override fun procurarPorId(id: UUID): Produto = repository.findById(id).orElseThrow { ProdutoNotFoundException() }

    override fun verificarQuantidade(id: UUID): Boolean = this.procurarPorId(id).quantidade!! > 0

    override fun procurar(): List<Produto> = repository.findAll()
}